### PR TITLE
core/types: add EffectiveGasPrice in Receipt

### PR DESCRIPTION
**Pull request by fjl**
Submitting this to take over PR #26250, which I cannot push to for some reason.

This change adds a struct field `EffectiveGasPrice` in `types.Receipt`. The field is present
in RPC responses, but not in the Go struct, and thus can't easily be accessed via ethclient.

### CODE REVIEW

**Review by holiman — APPROVED** [review]
LGTM (have not tested)

**Comment by fjl** [comment]
This should be ready for merging now!

File: core/types/tx_dynamic_fee.go [hunk 1]
```diff
@@ -94,6 +94,17 @@ func (tx *DynamicFeeTx) value() *big.Int        { return tx.Value }
 func (tx *DynamicFeeTx) nonce() uint64          { return tx.Nonce }
 func (tx *DynamicFeeTx) to() *common.Address    { return tx.To }
 
+func (tx *DynamicFeeTx) effectiveGasPrice(dst *big.Int, baseFee *big.Int) *big.Int {
+	if baseFee == nil {
+		return tx.GasFeeCap
```
[the remark this answers is not in the document]
Comment (holiman) [reply core/types/tx_dynamic_fee.go:99]: What does this mean, really? 
We want the 1559-transaction to tell us what the effective gas price is, if included in a block without a base fee. Is that what this clause means? 

And if so, is there really "A Correct Answer" to give in that situation? 

Secondly, it does seem more correct to use `dst.Set` as @MariusVanDerWijden points out, but I don't know if it should use `tx.GasPrice` or as `tx.GasFeeCap`

[the remark this answers is not in the document]
Comment (fjl) [reply core/types/tx_dynamic_fee.go]: Good catch! I have fixed it to set `GasFeeCap`. There is no `GasPrice` in `DynamicFeeTx`.

**Review by holiman — APPROVED** [review]
LGTM

File: core/types/tx_dynamic_fee.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (MariusVanDerWijden) [reply core/types/tx_dynamic_fee.go]: Ah sorry copy paste error, I meant gasFeecap

**Pull request merged**